src/Microsoft.VisualStudio.Threading/AsyncReaderWriterLock.cs

```diff
@@ -2160,7 +2160,21 @@ internal Awaiter(AsyncReaderWriterLock lck, LockKind kind, LockFlags options, Ca
             /// </summary>
             public bool IsCompleted
             {
-                get { return this.cancellationToken.IsCancellationRequested || this.fault != null || this.LockIssued; }
+                get
+                {
+                    if (this.fault != null)
+                    {
+                        return true;
+                    }
+
+                    // If lock has already been issued, we have to switch to the right context, and ignore the CancellationToken. 
+                    if (this.lck.IsLockActive(this, considerStaActive: true))
```

Comment:
if `LockIssued == true`, it seems we are already on the right context and requesting a nested lock. We don't seem to actually *issue* a lock until GetResult() or OnCompleted() is called. So I'm having a hard time seeing where the lock is leaked in this case.
I'm admittedly very rusty in this area.

Reply:
A lock can be issued in the constructor of the Awaitable object.  I didn't notice that earlier, and had hard time to find out the race condition too.  The problem happens when LockIssued == false, and cancellationToken.IsCancellationRequested == true.  Originally, IsCompleted will return true, and GetResult() will throw.